Allow passing null value to Uuid validators

jakobmerrild opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

## Description
Upgrading from 3.x to 4.x introduces a breaking change in the api of `Uuid::isValid` in that it will no longer accept `null` as input. 

With these changes `null` is now once again accepted as input to `Uuid::isValid` with the obvious result `false`.
## Motivation and context
We use your library extensively to check that various `string|null` values are valid uuids in our code base. While upgrading to `4.x` we noticed that some of these calls, namely those passing in `null`, would fail at runtime.

## How has this been tested?
I have not tested this beyond the additional test input in the relevant `XValidatorTest` files.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING.md** document.
- [x] I have added tests to cover my changes.

jakobmerrild committed:
Allow passing null value to Uuid validators

ramsey commented:
This changes the validator interface and introduces a BC break, so this would require a version 5.0 release.

Is there a reason you need to pass `null`? Can you check for `null` values before calling the validator?

jakobmerrild commented:
We can of course check for `null` prior to passing values to `Uuid::isValid`, but I didn't think this was a BC break since `string|null` is a super type of `string`. Am I wrong in this?

ramsey commented:
Changing the interface results in a fatal error like this:

```
Fatal error: Declaration of Foo::validate(string $uuid): bool must be compatible with
ValidatorInterface::validate(?string $uuid): bool in /in/Np3bf on line 10
```

See https://3v4l.org/Np3bf

jakobmerrild commented:
Of course! I didn't think of that 🙈.
I'm not familiar enough with PHPs type system to know for sure, but I assume that it isn't strong enough to accept `string|null` in the concrete class if the interface specifies only `string`?

ramsey commented:
We can give it a try... and it appears to work. 😄 

https://3v4l.org/lP0YG

jakobmerrild committed:
Revert "Allow passing null value to Uuid validators"

This reverts commit 7b44097248efcfe93e4847e999644cbebfef4f10.

jakobmerrild committed:
Allow null input to Uuid::isValid

Rather than updating the entire ValidatorInterface to allow
null input we simply change the static method in Uuid, which
is the method which had its api break when types were added
in version 4.

jakobmerrild commented:
I realised that `Uuid` does not seem to implement `ValidatorInterface`, so it's sufficient to simply move the null check to the static method `::isValid` in order to achieve the same interface as it had in version 3.

jakobmerrild committed:
Merge branch '4.x' into validator-allow-null-strings

jakobmerrild commented:
@ramsey Have you had a chance to look at the updated changes?

jakobmerrild commented:
@ramsey Just following up here. Wondering if this is something you'd be willing to include in the project?

ramsey committed:
Merge branch '4.x' into validator-allow-null-strings

ramsey commented:
I'm a little concerned about the change, since the Uuid class isn't `final`. Others might still extend Uuid, and if they reimplement `isValid()`, then they'll get a fatal error.

For example, see here: https://3v4l.org/SsOGK

jakobmerrild commented:
I didn't think of that either 🤔 .

Well it's definitely not worth introducing a breaking change in order to revert a breaking change. After all going from version 3 to 4 one would expect breaking changes 🙂 

I had a peak at the changelog. It doesn't seem to mention breaking changes in new major versions. Do you keep a list of these somewhere? Would then simply add to it that the `Uuid::isValid` no longer accepts `null` in version 4

ramsey commented:
What happened is that the docblock type for this method was always `string` and never `string | null`. In this commit, @garak added all the type hints according to the docblocks: https://github.com/ramsey/uuid/commit/5459113b816f4b1e15d30dffdc1d0e85bdaf4585#diff-898c45cfd882f04a4877c82e14ce28181f8ea6062654772bfb9fc474014bd6a0R686

In version 3.x, there are no type hints on the method, so it allowed anything, but the docblock specifies that it allows only `string`, so it was a bug that we accepted `null`, and this bug was fixed in the first alpha version of 4.0.0.

I'm sorry for the confusion.

jakobmerrild commented:
I understand. I guess we'll have to clean up our code base at some point 🙂 . For now we're pinning your library to `3.9`. 
Thanks for your time.

ramsey commented:
Sorry for the hassle. I just released 3.9.7, which fixes a few things so the library has no problem running on PHP 8.1 or 8.2.
